### PR TITLE
Change "method" to "associated function"

### PR DESCRIPTION
**Pull request by mark-i-m**
r? @matthewjasper 

cc @Centril @eddyb #67742 

I'm opening this mostly as a test to see what the diagnostic changes would be. It seems that this makes them somewhat more verbose, and I'm not sure it's worth it...

The relevant changes are the last two commits (it is rebased on top of #67742)

### CODE REVIEW

File: src/librustc/hir/map/mod.rs [hunk 1]
```diff
@@ -330,12 +330,12 @@ impl<'hir> Map<'hir> {
             },
             Node::TraitItem(item) => match item.kind {
                 TraitItemKind::Const(..) => DefKind::AssocConst,
-                TraitItemKind::Method(..) => DefKind::Method,
+                TraitItemKind::Method(..) => DefKind::AssocFn,
```
Comment (Centril) [inline src/librustc/hir/map/mod.rs]: We should also rename the `Method` variants to `Fn` as well.

**Comment by eddyb** [comment]
Can you open a PR to fix https://github.com/rust-lang/rust/issues/60163 without changing diagnostics?
Then we can separately do the diagnostic changes.

File: src/librustc_hir/def.rs [hunk 2]
```diff
@@ -107,7 +107,7 @@ impl DefKind {
             DefKind::Union => "union",
             DefKind::Trait => "trait",
             DefKind::ForeignTy => "foreign type",
-            DefKind::Method => "method",
+            DefKind::AssocFn => "associated function",
```
Comment (eddyb) [inline src/librustc_hir/def.rs]: For the record this is the same length as "associated constant" so it doesn't seem unnecessarily long to me.

Comment (Centril) [reply src/librustc_hir/def.rs]: Plus, it's always correct as opposed to sometimes misleading.

File: src/test/ui/suggestions/suggest-methods.stderr [hunk 3]
```diff
@@ -5,19 +5,19 @@ LL | struct Foo;
    | ----------- method `bat` not found for this
 ...
 LL |     f.bat(1.0);
-   |       ^^^ help: there is a method with a similar name: `bar`
+   |       ^^^ help: there is a associated function with a similar name: `bar`
```
Comment (Centril) [inline src/test/ui/suggestions/suggest-methods.stderr]:
```suggestion
   |       ^^^ help: there is an associated function with a similar name: `bar`
```

File: src/librustc/hir/map/mod.rs (hunk 1 above)
Comment (eddyb) [reply src/librustc/hir/map/mod.rs]: That's also included in #60163.

**Comment by bjorn3** [comment]
To me a `method` is a special kind of associated function which takes a `self` parameter.

**Comment by mark-i-m** [comment]
@bjorn3 I agree, but the compiler doesn't currently make a distinction iiuc

**Comment by bjorn3** [comment]
It would be nice if the error messages do make the distinction.

File: src/test/ui/suggestions/suggest-methods.stderr (hunk 3 above)
Comment (mark-i-m) [reply src/test/ui/suggestions/suggest-methods.stderr]: NOTE TO SELF: I will need to fix the article...

**Comment by mark-i-m** [comment]
Rebased over #69674 

@rustbot modify labels: -S-waiting-on-review +S-blocked

**Comment by Centril** [comment]
> It would be nice if the error messages do make the distinction.

I think error messages should only make the distinction where it actually matters. Otherwise, you have to scan for `self` in many places where it is semantically irrelevant whether `self` is there or not.

**Comment by matthewjasper** [comment]
@bors r+
cc @rust-lang/wg-diagnostics

**Comment by estebank** [comment]
>> It would be nice if the error messages do make the distinction.
>
> I think error messages should only make the distinction where it actually matters. Otherwise, you have to scan for self in many places where it is semantically irrelevant whether self is there or not.

I disagree, I think it makes it much easier to teach the concepts if we are consistent in calling associated functions with `self` parameters "methods". I've seen this confusion play out many times and it is our responsibility to reduce "induced" confusion.

**Comment by mark-i-m** [comment]
I guess I've personally experienced the opposite... in this case, I always used to here "associated X" and not really know what it meant because e.g. associated `fn` where inconsistently referred to as methods, functions, or assoc functions. Picking one and sticking to it would be good.

On the other hand, I think the main point in favor of `method` is that it is the terminology used by most other languages (e.g. java, c++).

**Comment by Centril** [comment]
Java also has the concept of "static method", which is our notion of an associated method without `self`, so that adds more to the confusion. Meanwhile, Haskell calls any function in a type class "method".

**Pull request merged**